compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -265,8 +265,13 @@ trait SpaceLogic {
       case (Prod(tp1, _, _, false), Typ(tp2, _)) =>
         if (isSubType(tp1, tp2)) Empty
         else a
-      case (Typ(tp1, _), Prod(tp2, _, _, false)) =>
-        a  // approximation
+      case (Typ(tp1, _), Prod(tp2, unappTp, params, false)) =>
+        if unappTp.symbol == defn.TypeTest_unapply then
+          // The `TypeTest[S, T].unapply` covers all `T`s if the scrutinee is of type `S`.
+          // If the scrutinee is not of type `S`, then we would already have an unchecked warning.
+          minus(a, params.head)
+        else
+          a  // approximation
```

Comment:
If the issues are really urgent to fix, I'd suggest extracting the code in #11186 (without introducing any annotations) and specialize the method `covers` instead.  This way,  we keep the logic in the base class general, which is a  design invariant supposed to improve maintainability of the code.

Reply:
@liufengyun could you help me with that?

Reply:
No problem, I'll create a new PR to do that today.